wasm-runtime/polkadot/src/runtime/governance.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Governance system: Handles administration and dispatch of sensitive operations including
+//! setting new code, minting new tokens and changing parameters.
+//!
+//! For now this is limited to a simple qualified majority vote (whose parameter is retrieved from
+//! storage) between validators. A single vote may be proposed per era, and at most one approval
+//! vote may be cast by each validator. The tally is maintained through a simple tag in storage for
+//! each validator that has approved.
+//!
+//! At the end of the era, all validators approvals are tallied and if there are sufficient to pass
+//! the proposal then it is enacted. All items in storage concerning the proposal are reset.
+
+use runtime_support::prelude::*;
+use keyedvec::KeyedVec;
+use storable::{Storable, StorageVec, kill};
+use primitives::{AccountID, Hash, BlockNumber};
+use proposal::Proposal;
+use runtime::{staking, system, session};
+
+// TRANSACTION API
+
+/// Propose a sensitive action to be taken. Any action that is enactable by `Proposal` is valid.
+/// Proposal is by the `transactor` and will automatically count as an approval. Transactor must
+/// be a current validator. It is illegal to propose when there is already a proposal in effect.
+pub fn propose(validator: &AccountID, proposal: &Proposal) {
+	if Proposal::lookup(b"gov:pro").is_some() {
+		panic!("there may only be one proposal per era.");
+	}
+	proposal.store(b"gov:pro");
+	approve(validator, staking::current_era());
+}
+
+/// Approve the current era's proposal. Transactor must be a validator. This may not be done more
+/// than once for any validator in an era.
+pub fn approve(validator: &AccountID, era_index: BlockNumber) {
+	if era_index != staking::current_era() {
+		panic!("approval vote applied on non-current era.")
```

Comment:
is panic being used as _the_ signal for block processing failure? seems dangerous if we want to run it in a native context.

It would be better to propagate errors up using something like `error_chain` and use the `bail!` macro.

Reply:
disagree vehemently. that would add a tonne of noise for no reason. within the context of the wasm runtime, a panic is always tantamount to block failure. native has a panic handler. unless there are cirsumstances where an intentional panic in the runtime would cause the main node to panic, then i see no reason to add it.